src/k8s/cmd/k8sd/k8sd_cluster_recover.go

```diff
@@ -0,0 +1,303 @@
+	m, err := microcluster.App(
+		microcluster.Args{
+			StateDir: rootCmdOpts.stateDir,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	// The following method parses cluster.yaml and filters out the entries
+	// that are not included in the trust store. Note that in case of k8s-dqlite,
+	// there is no trust store.
+	members, err := m.GetDqliteClusterMembers()
+	if err != nil {
+		return fmt.Errorf("Could not retrieve K8sd cluster members: %w.", err);
+	}
+
+	oldMembersYaml, err := yaml.Marshal(members)
+	if err != nil {
+		return err
+	}
+
+	clusterYamlPath := path.Join(m.FileSystem.DatabaseDir, "cluster.yaml")
+	clusterYamlCommentHeader := fmt.Sprintf("# K8sd cluster configuration\n# (%s)\n", clusterYamlPath)
+	// We aren't passing the file path to our text editor instance since we don't want
+	// the actual content to be written before we get to validate it.
+	clusterYamlContent, err := shared.TextEditor(
+		"",
+		slices.Concat(
+			[]byte(clusterYamlCommentHeader),
+			[]byte(clusterYamlRecoveryComment),
+			oldMembersYaml))
+	if err != nil {
+		return err
+	}
+
+	infoYamlPath := path.Join(m.FileSystem.DatabaseDir, "info.yaml")
+	infoYamlCommentHeader := fmt.Sprintf("# K8sd info.yaml\n# (%s)\n", infoYamlPath)
+	_, err = shared.TextEditor(
+		"",
+		slices.Concat(
+			[]byte(infoYamlCommentHeader),
+			[]byte(infoYamlRecoveryComment),
+			yamlCommentLines(removeYamlComments(clusterYamlContent))))
+	if err != nil {
+		return err
+	}
+	err = removeYamlCommentsFromFile(infoYamlPath)
+	if err != nil {
+		return err
+	}
+
+	daemonYamlPath := path.Join(m.FileSystem.StateDir, "daemon.yaml")
+	daemonYamlCommentHeader := fmt.Sprintf("# K8sd daemon.yaml\n# (%s)\n", daemonYamlPath)
+	_, err = shared.TextEditor(
+		"",
+		slices.Concat(
+			[]byte(daemonYamlCommentHeader),
+			[]byte(daemonYamlRecoveryComment),
+			yamlCommentLines(removeYamlComments(clusterYamlContent))))
+	if err != nil {
+		return err
+	}
+	err = removeYamlCommentsFromFile(daemonYamlPath)
+	if err != nil {
+		return err
+	}
+
+	newMembers := []cluster.DqliteMember{}
+	err = yaml.Unmarshal(clusterYamlContent, &newMembers)
+	if err != nil {
+		return err
+	}
+
+	// The following microcluster method will:
+	// * validate the member changes
+	//     * ensure that no members were added or removed
+	//     * the member IDs hasn't changed
+	//     * there is at least one voter
+	//     * the addresses can be parsed
+	//     * there are no duplicate addresses
+	// * ensure that all cluster members are stopped
+	// * create a database backup
+	// * reconfigure Raft
+	// * address changes based on the new cluster.yaml:
+	//     * refresh the local info.yaml and daemon.yaml
+	//     * update the trust store addresses
+	//     * prepare an sql script that updates the member addresses from the
+	//       "core_cluster_members" table, executed when k8sd starts.
+	// * rewrite cluster.yaml
+	// * create a recovery tarball of the k8sd database dir and store it
+	//   in the state dir.
```

Comment:
As an alternative, I could say ``As of 2.0.2, the following microcluster will``.

I was a bit concerned that the comment would go out of date, but IMO it helps understand which checks and backup actions are performed by the microcluster method and which should be performed by us.

Otherwise, someone reading the code may find it confusing that we're checking the cluster members and peforming a backup for k8s-dqlite and not for k8sd, the reason being that the latter is handled by microcluster.

Reply:
makes sense